go/extractor/diagnostics/diagnostics.go

```diff
@@ -0,0 +1,181 @@
+	CliSummaryTable bool `json:"cliSummaryTable"` // True if the message should be counted in the diagnostics summary table printed by `codeql database analyze` (defaults to false)
+	Telemetry       bool `json:"telemetry"`       // True if the message should be sent to telemetry (defaults to false)
+}
+
+var fullVisibility *visibilityStruct = &visibilityStruct{true, true, true}
+
+type locationStruct struct {
+	File        string `json:"file,omitempty"`
+	StartLine   int    `json:"startLine,omitempty"`
+	StartColumn int    `json:"startColumn,omitempty"`
+	EndLine     int    `json:"endLine,omitempty"`
+	EndColumn   int    `json:"endColumn,omitempty"`
+}
+
+var noLocation *locationStruct = nil
+
+type diagnostic struct {
+	Timestamp       string            `json:"timestamp"`
+	Source          sourceStruct      `json:"source"`
+	MarkdownMessage string            `json:"markdownMessage"`
+	Severity        string            `json:"severity"`
+	Visibility      *visibilityStruct `json:"visibility,omitempty"` // Use a pointer so that it is omitted if nil
+	Location        *locationStruct   `json:"location,omitempty"`   // Use a pointer so that it is omitted if nil
+}
+
+var diagnosticsEmitted, diagnosticsLimit uint = 0, 100
+var noDiagnosticDirPrinted bool = false
+
+func emitDiagnostic(sourceid, sourcename, markdownMessage string, severity diagnosticSeverity, visibility *visibilityStruct, location *locationStruct) {
+	if diagnosticsEmitted < diagnosticsLimit {
+		diagnosticsEmitted += 1
+
+		diagnosticDir := os.Getenv("CODEQL_EXTRACTOR_GO_DIAGNOSTIC_DIR")
+		if diagnosticDir == "" {
+			if !noDiagnosticDirPrinted {
+				log.Println("No diagnostic directory set, so not emitting diagnostic")
+				noDiagnosticDirPrinted = true
+			}
+			return
+		}
+
+		timestamp := time.Now().UTC().Format("2006-01-02T15:04:05.000") + "Z"
+
+		var d diagnostic
+
+		if diagnosticsEmitted < diagnosticsLimit {
+			d = diagnostic{
+				timestamp,
+				sourceStruct{sourceid, sourcename, "go"},
+				markdownMessage,
+				string(severity),
+				visibility,
+				location,
+			}
+		} else {
+			d = diagnostic{
+				timestamp,
+				sourceStruct{"go/autobuilder/diagnostic-limit-reached", "Diagnostics limit exceeded", "go"},
+				fmt.Sprintf("CodeQL has produced more than the maximum number of diagnostics. Only the first %d have been reported.", diagnosticsLimit),
+				string(severityWarning),
+				fullVisibility,
+				noLocation,
+			}
+		}
+
+		content, err := json.Marshal(d)
+		if err != nil {
+			log.Println(err)
+			return
+		}
+
+		targetFile, err := os.CreateTemp(diagnosticDir, "go-extractor.*.json")
+		if err != nil {
+			log.Println("Failed to create temporary file for diagnostic: ")
+			log.Println(err)
+		}
+		defer func() {
+			if err := targetFile.Close(); err != nil {
+				log.Println("Failed to close diagnostic file:")
+				log.Println(err)
+			}
+		}()
+
+		_, err = targetFile.Write(content)
+		if err != nil {
+			log.Fatal(err)
```

Comment:
Remove fatal / panic exits from diagnostic code -- there's no reason to bring the whole process down if diagnostic writing fails